internal/engine/upper.go

```diff
@@ -52,138 +51,53 @@ func NewUpper(ctx context.Context, manager service.Manager, env k8s.Env) (Upper,
-
-	}()
-	return ret
-}
-
 func (u Upper) Up(ctx context.Context, services []model.Service, watchMounts bool) error {
 	span, ctx := opentracing.StartSpanFromContext(ctx, "daemon-Up")
 	defer span.Finish()
-	var buildTokens []*buildToken
-	for i := range services {
-		buildToken, err := u.b.BuildAndDeploy(ctx, services[i], nil, nil)
-		buildTokens = append(buildTokens, buildToken)
-		if err != nil {
-			return err
-		}
-	}
 
+	buildTokens := make(map[model.ServiceName]*buildToken)
+
+	var sw *serviceWatcher
+	var err error
 	if watchMounts {
-		service := services[0]
-		if len(services) > 1 {
-			return errors.New("There is more than 1 service")
-		}
-		watcher, err := u.watcherMaker()
+		sw, err = makeServiceWatcher(ctx, u.watcherMaker, u.timerMaker, services)
 		if err != nil {
 			return err
 		}
+	}
 
-		if len(service.Mounts) == 0 {
-			return errors.New("service has 0 repos - nothing to watch")
-		}
-
-		var repoRoots []string
-
-		for _, mount := range service.Mounts {
-			repoRoots = append(repoRoots, mount.Repo.LocalPath)
-			err = watcher.Add(mount.Repo.LocalPath)
-			if err != nil {
-				return err
-			}
-		}
-
-		eventFilter, err := git.NewMultiRepoIgnoreTester(ctx, repoRoots)
+	for _, service := range services {
+		buildToken, err := u.b.BuildAndDeploy(ctx, service, nil, nil)
+		buildTokens[service.Name] = buildToken
 		if err != nil {
 			return err
 		}
+	}
 
-		coalescedEvents := u.coalesceEvents(watcher.Events())
-
+	if watchMounts {
 		for {
 			select {
-			case err := <-watcher.Errors():
-				return err
-			case events, ok := <-coalescedEvents:
-				if !ok {
-					return nil
-				}
-				logger.Get(ctx).Infof("files changed, rebuilding %v", service.Name)
-
-				var changedPaths []string
-				for _, e := range events {
-					path, err := filepath.Abs(e.Name)
-					if err != nil {
-						return err
-					}
-					isIgnored, err := eventFilter.IsIgnored(path, false)
-					if err != nil {
-						return err
-					}
-					if !isIgnored {
-						changedPaths = append(changedPaths, path)
-					}
-				}
-				if len(changedPaths) > 0 {
-					var changedPathsToPrint []string
-					if len(changedPaths) > maxChangedFilesToPrint {
-						changedPathsToPrint = append(changedPaths[:maxChangedFilesToPrint], "...")
-					} else {
-						changedPathsToPrint = changedPaths
-					}
-					logger.Get(ctx).Infof("files changed. rebuilding %v. observed changes: %v", service.Name, changedPathsToPrint)
-
-					buildTokens[0], err = u.b.BuildAndDeploy(ctx, service, buildTokens[0], changedPaths)
+			case event := <-sw.events:
+				var changedPathsToPrint []string
+				if len(event.files) > maxChangedFilesToPrint {
+					changedPathsToPrint = append(event.files[:maxChangedFilesToPrint], "...")
+				} else {
+					changedPathsToPrint = event.files
 				}
+
+				logger.Get(ctx).Infof("files changed. rebuilding %v. observed changes: %v", event.service.Name, changedPathsToPrint)
+
+				var err error
+				buildTokens[event.service.Name], err = u.b.BuildAndDeploy(
```

Comment:
if there's a build error, you probably don't want to overwrite the old token?

Reply:
yeah, that was unintentional. thanks.

(though theoretically even a failed build could generate artifacts that would speed up successive builds and maybe the contract should be we always returns a token, even if it's just the one that was passed in?)